src/v/archival/ntp_archiver_service.cc

```diff
@@ -854,4 +872,136 @@ ntp_archiver::maybe_truncate_manifest(retry_chain_node& rtc) {
+    if (!retention_calculator) {
+        co_return;
+    }
+
+    auto next_start_offset = retention_calculator->next_start_offset();
+    if (next_start_offset) {
+        vlog(
+          _rtclog.debug,
+          "{} Advancing start offset to {} satisfy retention policy",
+          retention_calculator->strategy_name(),
+          *next_start_offset);
+
+        auto sync_timeout = config::shard_local_cfg()
+                              .cloud_storage_metadata_sync_timeout_ms.value();
+        auto deadline = ss::lowres_clock::now() + sync_timeout;
+        auto error = co_await _partition->archival_meta_stm()->truncate(
+          *next_start_offset, deadline, _as);
+        if (error != cluster::errc::success) {
+            vlog(
+              _rtclog.warn,
+              "Failed to update archival metadata STM start offest according "
+              "to retention policy: {}",
+              error);
+        }
+    } else {
+        vlog(
+          _rtclog.debug,
+          "{} Retention policies are already met.",
+          retention_calculator->strategy_name());
+    }
+}
+
+ss::future<> ntp_archiver::garbage_collect() {
+    if (!housekeeping_can_continue()) {
+        co_return;
+    }
+
+    std::vector<remote_segment_path> to_remove;
+    auto current_start_offset = manifest().get_start_offset();
+
+    for (const auto& [key, value] : manifest()) {
+        if (value.committed_offset < current_start_offset) {
+            to_remove.emplace_back(
+              manifest().generate_remote_segment_path(_ntp, key, value));
+        }
+    }
+
+    auto replaced = manifest().replaced_segments();
+    for (const auto& meta : replaced) {
+        const auto segment_name = cloud_storage::generate_local_segment_name(
+          meta.base_offset, meta.segment_term);
+        to_remove.emplace_back(cloud_storage::generate_remote_segment_path(
+          _ntp, meta.ntp_revision, segment_name, meta.archiver_term));
+    }
+
+    // TODO(vlad): Maybe make this loop concurrent.
+    auto deletions_failed = false;
+    for (const auto& path : to_remove) {
+        retry_chain_node fib(
+          _metadata_sync_timeout, _cloud_storage_initial_backoff, &_rtcnode);
+        retry_chain_logger ctxlog(archival_log, fib, _ntp.path());
+
+        // If deletion fails and the subsequent command to cleanup metadata
+        // succeeds, the segment will become orphaned.
+        auto res = co_await _remote.delete_object(
+          _bucket, s3::object_key{path}, fib);
+        if (res != cloud_storage::upload_result::success) {
+            deletions_failed = true;
+            break;
+        }
+    }
+
+    if (!deletions_failed) {
+        auto sync_timeout = config::shard_local_cfg()
+                              .cloud_storage_metadata_sync_timeout_ms.value();
+        auto deadline = ss::lowres_clock::now() + sync_timeout;
+        auto error = co_await _partition->archival_meta_stm()->cleanup_metadata(
```

Comment:
I think that there is a race condition here. Between the initialization of the `to_remove` collection and the call to `cleanup_metadata` there are a lot of async calls and it can take awhile to delete everything in `to_remove`. If during deletion the `start_offset` will be moved forward or some segments will be replaced, the metadata of those segments will be lost. I think that the garbage collection should block uploads. Or alternatively, it should be called from the upload loop. The houskeeping loop could arrange that somehow (with the flag or a semaphore).

Reply:
I can see the merit in having a separate fiber that runs retention especially when we want upload and retention to have different priorities.

Reply:
The `start_offset` is only advanced by house keeping so that should not be an issue. That leaves the reuploading of compacted segments which is indeed an issue. Could we just block the re-upload of compacted segments while GC is running? This leaves normal uploads untouched.

Reply:
I would be +1 on making retention updates, uploads and re-uploads all mutually-exclusive in time.  Makes it so much easier to reason about and test.

Reply:
Done. I moved the housekeeping into the upload loop. I feel like this is a bit of a big hammer. We could have just blocked compacted uploads, but let's come back to it after the release.